### PR TITLE
Refactor `AlignAndFocusPowderFromFiles` to save and load summed cache 

### LINKED ISSUE #24075: Refactor `AlignAndFocusPowderFromFiles` to save and load summed cache files
When a list of runs is provided to `AlignAndFocusPowderFromFiles`, we create a cached file for the summed runs, not just individual cache files per run (current behavior). Also, have the mechanism in-place to find the summed cached files for (in this order)
1. the summed cache file for all listed runs
2. a summed cache file for a subset of runs listed
3. individual cache files for the runs

We save and load individual cache files. This is slow (at least for `NOMAD` cache files) for summing many runs. This would allow for directly loading the (summed) cache files which should be significantly faster.

The loop to refactor is found starting here: https://github.com/mantidproject/mantid/blob/fdce3b20d7d843e44ffca31d25daea62bd71de2b/Framework/PythonInterface/plugins/algorithms/AlignAndFocusPowderFromFiles.py#L304

### PR DESCRIPTION
**Pull request by yxqd**
**Description of work.**

* Create caches when computing a sum, including caches of partial sums
* Use cache in computing sum by using a greedy algorithm to find caches

**Report to:** 
@marshallmcdonnell  

**To test:**

Fixes #24075 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by VickieLynch — COMMENTED** [review]
Saved results from using these cached files and compared to results from code in master and they agree using CheckWorkspacesMatch and looking at plots.

**Review by VickieLynch — COMMENTED** [review]
Running master with cached files took 71.6 seconds.  Running this with cached sums took 14 seconds.

**Review by marshallmcdonnell — DISMISSED** [review]
Similarly, I had the following results running 4 NOMAD files for the following cases:
1. initial (no cacheing): `531. secs`
2. summed cache file: `27. secs`
3. summed cache file + one more run (4 cached in sum + 1 uncached) :  `406. secs`

Naive performance analysis:
  Non-cached expected value: `663.75 secs`  = `531. + 531/4.` 
  Summed + one uncached expected value: `159.75 secs`
  Actual summed + one uncached value: `440. secs`
  Difference: `280.25 secs` = `440. - ( 27. + 531./4.)`

So it looks like the `280.25 secs` is what to expect for overhead to add an uncached file (summing the uncached files with summed, etc.) Definitely a performance boost, especially for the summed cache file!

Approve for merge.

**Comment by marshallmcdonnell** [comment]
@gvardany agreed, added release note with previous note about improvements for chunking.

**Review by marshallmcdonnell — DISMISSED** [review]

**Comment by yxqd** [comment]
Please don't merge yet. I am experimenting a change to improve memory footprint.

**Comment by yxqd** [comment]
@marshallmcdonnell for your test of 5 files: 4 with cache, 1 without cache, did you try to run the algorithm with one single file? I wonder if that particular file just takes longer.

File: Framework/PythonInterface/plugins/algorithms/AlignAndFocusPowderFromFiles.py
Comment (peterfpeterson) [inline Framework/PythonInterface/plugins/algorithms/AlignAndFocusPowderFromFiles.py]: Don't you need to either delete the "grain" or reset it?

Comment (yxqd) [reply Framework/PythonInterface/plugins/algorithms/AlignAndFocusPowderFromFiles.py]: Each "grain" is named differently. So no reset is needed. And when it is "accumulated" to the final sum, it is deleted

File: Framework/PythonInterface/plugins/algorithms/AlignAndFocusPowderFromFiles.py
Comment (peterfpeterson) [inline Framework/PythonInterface/plugins/algorithms/AlignAndFocusPowderFromFiles.py]:
```suggestion
                # accumulate into final sum
```

**Review by marshallmcdonnell — APPROVED** [review]
Additional testing showed more benefit to memory. Overall, great improvement that will really have an effect on NOMAD reduction (and diffractometers using the algorithm) :+1: 

I approve for merge (if build CI passes)

**Pull request merged**
